pkg/sentry/loader/elf.go

```diff
@@ -488,6 +488,17 @@ func loadParsedELF(ctx context.Context, m *mm.MemoryManager, f *fs.File, info el
 			panic(fmt.Sprintf("Failed to unmap base address: %v", err))
 		}
 
+		// Offset is used as start virtual address of shared object, the
+		// map and unmap will ensure there are enouth space for all
+		// PT_LOAD phdrs.
+		// Later using of offset as virtual address 0 of the share object,
+		// this assume is right if the virtual address of phdrs start at 0.
+		// But this is not true if prelink is enabled, maybe some other
+		// reasons, this will make the mmap in mapSegment failed unexpected.
+		if offset != 0 {
+			offset = offset - usermem.Addr(start).RoundDown()
```

Comment:
This seems incorrect to me. If I understand correctly, this is effectively cancelling out the addition immediately below, such that we will in face map at the address returned by MMap above.

It seems that prelink(3) expects libraries to actually be loaded at the fixed vaddr. It does seem to maintain relocations such that the dynamic linker can still handle the library even if it is loaded at a different address, but that shouldn't be the common case.

I haven't fully grokked the kernel code, but it looks like we should be using vaddr as the hint address in mapping for interpreters (but not the main binary). I think we need more complete tests to verify the behavior on Linux.